kythe/docs/rfc/3339.md

```diff
@@ -0,0 +1,391 @@
+    receives output from the analyzer to the driver. In a POSIX environment,
+    these streams correspond to stdin and stdout.
+
+-   The data on each channel are partitioned into consecutive **frames**. Each
+    frame consists of a length tag followed by a message body of that length.
+
+    A **length tag** is a byte count _n_ encoded as a minimum-width decimal
+    integer using the Unicode digits `0` (code 48) through `9` (code 57), and
+    terminated by a newline (code 10). A **message body** is the sequence of _n_
+    bytes immediately following the length tag.
+
+    For example, a frame containing the message `foobar` is encoded as
+
+        6    \n  f    o    o    b    a    r      -- text
+        0x36 0xa 0x66 0x6f 0x6f 0x62 0x61 0x72   -- Unicode (hex)
+
+-   Within this framing, the analyzer communicates with the driver using
+    [JSON-RPC 2.0](http://www.jsonrpc.org/specification) requests and responses,
+    as defined below. The analyzer is the JSON-RPC client, and the driver is the
+    server.
+
+    The server MUST support concurrent JSON-RPC requests from the analyzer. The
+    server MUST support batch JSON-RPC requests from the analyzer.
+
+-   Under certain circumstances, the analyzer process **must exit**. In a case
+    where the analyzer process must exit, it MUST terminate within a grace
+    period. If it does not, the operating environment SHOULD forcibly terminate
+    the process (and may do so at any time after the grace period ends). The
+    exit code of an analyzer process is ignored. The grace period is defined by
+    the implementation, but must be at least 10 seconds.
+
+-   If and after the analyzer closes the output stream, the analyzer process
+    must exit.
+
+-   If and after the driver closes the input stream, the analyzer process must
+    exit.
+
+-   A **corrupt frame** is one in which the length tag does not have the correct
+    form. A frame is corrupt if the newline terminator is missing, the string
+    preceding the newline terminator contains non-digits or is empty, or the
+    stream ends before the specified number of bytes is delivered.
+
+    If the analyzer receives a corrupt frame from the driver, it must exit. If
+    the driver receives a corrupt frame from the analyzer, it must close the
+    input stream (in which case the analyzer therefore must exit).
+
+-   When an analyzer process terminates, the driver MUST consider any pending
+    analysis requests to have failed, and thereafter the environment MAY retry
+    analysis of the corresponding compilation records.
+
+-   If the analyzer voluntarily (i.e. not at behest of the driver) exits
+    succesfully (e.g. exit code of 0) with no outstanding analyses, the driver
+    SHOULD cause a new instance of the analyzer to start. This is motivated by
+    analyzers for which the best, or only, mechanism for clearing state is to
+    restart.
+
+### Messages
```

Comment:
(aside: it would be nice to specify the protocol using proto, even if the implementation is JSON-RPC)

Reply:
Yeah, that's what kythe/proto/driver.proto was meant to represent, but I think it may be out-of-date with the current proposal.